BIP 181, 182, 183: BIPs for Utreexo (#1923)
bitcoin/bips: bip-0182.md

```
  BIP: 182
  Layer: Peer Services
  Title: Utreexo - Transaction and block validation
  Author: Tadge Dryja <rx@awsomnet.org>
          Calvin Kim <bip@calvinkim.info>
          Davidson Souza <bip@dlsouza.dev>
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0182
  Status: Draft
  Type: Standards Track
  Created: 2023-10-01
  License: BSD-3-Clause

>   Requires: 181

Reviewer: ```suggestion
  BIP: 182
  Layer: Peer Services
  Title: Utreexo - Transaction and block validation
  Authors: Tadge Dryja <rx@awsomnet.org>
           Calvin Kim <bip@calvinkim.info>
           Davidson Souza <bip@dlsouza.dev>
  Status: Draft
  Type: Specification
  Assigned: 2025-08-29
  License: BSD-3-Clause
  Requires: 181
```

```

## Abstract

This BIP defines the rules for validating blocks and transactions using the
Utreexo accumulator. It is important to note that this BIP does not define the
Utreexo accumulator itself, for that see [Utreexo Accumulator Specification](bip-0181.md). This document is only concerned with

> the general rules for validating blocks and transactions using the Utreexo,

Reviewer 2: ```suggestion
the general rules for validating blocks and transactions using Utreexo,
```

represent the UTXO set:

 - 1: The serialization format of each UTXO ("leaf data").
 - 2: The hash function used to hash the leaf data.
 - 3: Which transaction outputs are excluded from the accumulator.
 - 4: The order of operations for the additions and deletions in the accumulator.
 - 5: The format of the UTXO proof.

A discrepancy in any of the five components above will result in a divergent
accumulator state, leading to consensus incompatibilities.

## License

This BIP is licensed under the BSD 3-clause license.

## Specification

### Node Hashes

During a node's normal operation, it will need to compute the leaf hash for UTXOs
being added or removed from the accumulator. The leaf hash is a 32-byte hash that
is computed using the SHA-512/256 hash function. See [UTXO Hash Preimages](#utxo-hash-preimages) for the
details on how to compute the leaf hash.

Unless otherwise specified, all fields are in little-endian format.

#### UTXO Hash Preimages

Individual UTXOs are represented as 32-byte hashes in the Utreexo accumulator. To obtain this
hash, you must compute the SHA-512/256 hash of the following data:

> | Name               | Type                     | Description                               |

Reviewer 2: Inconsistent use of terminating periods in the Description column.

| ------------------ | ------------------------ | ----------------------------------------- |
| Utreexo_Tag_V1     | 64-byte array            | The version tag to be prepended to the leafhash. |

> | Utreexo_Tag_V1     | 64-byte array            | The version tag to be prepended to the leafhash. |

Reviewer 2: Duplicate table rows.

Reviewer 2: Sorry, didn't notice this was intentional. You could consolidate this into one row that says "128-byte array", or change the description of the second line to clarify the intended duplication.

| BlockHash          | 32-byte array            | The hash of the block in which this tx was confirmed. |
| TXID               | 32-byte array            | The transaction's TXID                    |
| Vout               | 4-byte unsigned integer  | The output index of this UTXO             |

> | Header code        | 4-byte unsigned integer  | The block height and iscoinbase. This is a value obtained by left shifting the block height that confirmed this transaction by one bit, and then OR-ing it with 1, only if this transaction is a coinbase. |

Reviewer 2: ```suggestion
| Header code        | 4-byte unsigned integer  | The block height and iscoinbase. This is a value obtained by left shifting the block height that confirmed this transaction by one bit, and then OR-ing it with 1 if this transaction is a coinbase. |
```

| Amount             | 8-byte unsigned integer  | The amount in satoshis for this UTXO      |

> | Output script size | varint                   | The output script length in bytes              |

Reviewer 2: "varint" is very often confused with "compact size", are you sure you didn't mean the latter? In any case you might want to link to some documentation for this format, as it might not be universally understood.

See: https://learnmeabitcoin.com/technical/general/compact-size/#varint

| Output script      | variable byte array      | The output script of the UTXO            |

Each field being defined as follows:

##### Version tag

We use [tagged hashes](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#user-content-Design) for the hashes committed in the accumulator for versioning
purposes. This is added so that if there are changes in the preimage of the
hash, the version tag helps to avoid misinterpretation.

The Utreexo version tag is the SHA512 hash of the string `UtreexoV1`, which is represented as the vector
`[85 116 114 101 101 120 111 86 49]` and hex `0x5574726565786f5631`.  (The resulting 64-byte output is

> `5b832db8ca26c25be1c542d6cceddda8c145615cff5c35727fb3462610807e20ae534dc3f64299199931772e03787d18156eb3151e0ed1b3098bdc8445861885`).

Reviewer 2: Inconsistent use of `0x` before a hexadecimal string.

##### Blockhash

We commit to the hash of the block which confirms the UTXO. This
is not currently used in the validation code, but could be used at a future
version to increase the work required for collision attacks.
A valid blockhash requires a large amount of work, which would prevent an
attacker from performing a standard cycle-finding collision attack in $2^{n/2}$
operations for an n-bit hash.

This could allow a later or alternate version to use shorter truncated hashes,
saving bandwidth and storage while still keeping Bitcoin's $2^{128}$ security.

##### TXID

The TXID is the transaction ID of the transaction that created this UTXO.

##### VOUT

The output index of the UTXO in the transaction.

##### Header code

> This field stores the block height and a boolean for marking that the UTXO was

Reviewer 2: ```suggestion
This field stores the block height and a boolean for marking whether the UTXO was
```

length.

##### Output script

This field is added to commit to the output script of the UTXO. With current
nodes, this is stored in the UTXO set but since we receive this in the proof
from our peers, we need to commit to this value to avoid malicious peers that
may send over the wrong output script.

#### Hash function

The leaf data is hashed with SHA-512/256, which gives us a 32 byte hash.
It was chosen over SHA-256 due to the faster performance on 64 bit systems.

#### Excluded UTXOs from the accumulator

Not all transaction outputs are added to a node's UTXO set. Normal Bitcoin nodes
only form consensus on the set of transactions, not on the UTXO set, so different
nodes can omit different outputs and stay compatible as long as those outputs are
never spent. Utreexo nodes, however, do require explicit consensus on the UTXO set
as all proofs are with respect to the Merkle roots of the entire set.

For this reason, we define which UTXOs are not inserted to the accumulator.  Any
variations here will result in Utreexo nodes with incompatible proofs.

##### Provably unspendable transaction outputs

> There are outputs in the Bitcoin network that we can guarantee that they cannot

Reviewer 2: ```suggestion
There are outputs in the Bitcoin network we can guarantee cannot
```

Often, UTXOs are created and spent in the same block. This is allowed by Bitcoin
consensus rules as long as the output being spent is created by a transaction earlier
in the block than the spending transaction.
In Utreexo, nodes inspect blocks and identify which outputs are being created
and destroyed in the same block, and exclude them from the accumulator and proofs.

There's no need to provide proofs for outputs which have been created in the same
block. Adding and then immediately removing the output from the accumulator would be
possible but doesn't serve any purpose - once outputs are spent, their past existence
cannot be proven with the Utreexo accumulator (and SPV proofs already provide that).

For these reasons, outputs which are spent in the same block where they are created
are omitted from the accumulator, and those inputs are omitted from block proofs.

#### Order of operations

The Utreexo accumulator lacks associative properties during addition and the
ordering of which UTXO hash gets added first is consensus critical. For
the modification of the accumulator the steps are as follows:

1. Batch remove the UTXOs that were spent in the block based on the algorithm

>    defined in [Utreexo Accumulator Specification](bip-0181.md). Deletions itself are order-independent.

Reviewer 2: ```suggestion
   defined in [Utreexo Accumulator Specification](bip-0181.md). Deletions themselves are order-independent.
```

2. Batch add all non-excluded outputs in the order they're included in the
   Bitcoin block. Additions are order-dependent.

The removal and the addition of the hashes follow the algorithms defined in
[Utreexo Accumulator Specification](bip-0181.md).

#### Format of the UTXO proof

The UTXO proof has 2 elements: the accumulator proof and the leaf data. The
leaf data provides the necessary UTXO data for block validation that would be
stored locally for non-Utreexo nodes. Non-Utreexo nodes store this data (under "chainstate/" for Bitcoin Core)

> but since utreexo nodes don't this data, it must be provided.

Reviewer: Missing a word.

```suggestion
but since utreexo nodes don't <missing word> this data, it must be provided.
```

> but since utreexo nodes don't this data, it must be provided.

Reviewer 2: ```suggestion
but since Utreexo nodes don't this data, it must be provided.
```

> The accompanying accumulator proof proves that the given leaf data are committed

Reviewer 3: ```suggestion
The accompanying accumulator proof proves that the given leaf data items are committed
```

in the accumulator. Without this accumulator proof, the Utreexo nodes would not have
a way to ensure that the given UTXO data exists in the UTXO set. The accumulator proof
and the verification ensures that Utreexo nodes are not fooled into accepting transactions
whose input UTXOs do not exist.

[Accumulator Proof](bip-0181.md#definitions) is defined in BIP-0181, and contains two elements:

1. A vector of positions of the UTXO hashes in the accumulator.
2. A vector of hashes required to hash up to the roots.

For (1), positions are in the order of the leaves that are being proven in
the accumulator. These are all the inputs in the natural blockchain order that
excludes the same block spends.

The UTXO hash preimages follow the same ordering as (1) in the accumulator
proofs. Each of the positions in (1) refer to the UTXO hash preimage in the same
index.

| Field Name          | Data Type           | Byte Size | Description                              |
| ------------------- | ------------------- | --------- | ---------------------------------------- |
| Accumulator Proof   | variable byte array | variable  | The Utreexo proof as defined in BIP-0181 |

> | UTXO hash preimages | variable byte array | variable  | The UTXO data needed to validate all the transaction in the block |

Reviewer 3: ```suggestion
| UTXO hash preimages | variable byte array | variable  | The UTXO data needed to validate all the transactions in the block |
```

#### UTXO proof validation

> For each block, the UTXO proof must be provided with the bitcoin block for

Reviewer 2: ```suggestion
For each block, the UTXO proof must be provided with the Bitcoin block for
```

validation to be possible. Without the UTXO proof, it's not possible to

> validate that the inputs being referenced exists in the UTXO set.

Reviewer 2: ```suggestion
validate that the inputs being referenced exist in the UTXO set.
```

hash preimages that are required to perform the rest of the consensus
validation checks. Note that the resulting data from the UTXO proof validation
is the same data that would normally be fetched from the locally stored UTXO
set.

The order of operations for the UTXO proof validation are:

1. Hash the UTXO preimages.
2. Verify that the UTXO preimages exist in the accumulator with the verification
   algorithm specified in BIP-0181.

### BIP-0030

[`BIP-0030`](https://github.com/bitcoin/bips/blob/master/bip-0030.mediawiki) is an added
consensus check that prevents duplicate TXIDs. This check and the historical violations
of this check affect the consensus validation for Utreexo nodes.

### BIP-0030 and BIP-0034 consensus check

Before `BIP-0030`, the Bitcoin consensus rules allowed for duplicate TXIDs. If two
transactions shared a same TXID, the transaction outputs of the succeeding
transaction would overwrite the previously created UTXOs. It was assumed that
TXIDs were unique but it was trivially easy to create a duplicate transaction that was
exactly the same, resulting in a duplicate `TXID` for coinbase transactions by re-using

> the same bitcoin address.

Reviewer 2: ```suggestion
the same Bitcoin address.
```

`BIP-0030` check is a consensus check that enforces that newly created transactions
do not have outputs that overwrite an existing UTXO.

`BIP-0034` introduces a rule that requires the block height to be included in the coinbase field
of the coinbase transaction. The main reason for the change was to make
coinbase transactions unique so that the expensive check of going through the
UTXO set wouldn't be needed. However, there were blocks in the past that had

> random bytes that could be interpreted as block heights. The lowest implicated block

Reviewer 2: ```suggestion
random bytes that could be interpreted as block heights. The lowest impacted block
```

heights are: 209,921, 490,897, and 1,983,702.

Up until block 209,921 the BIP-0030 checks are performed for non-Utreexo nodes.
Since Utreexo nodes only keep the UTXO set commitment, it's not possible to
perform the `BIP-0030` check. In theory, those blocks can't be reorged, because
of checkpoints, that goes back to block height 295,000 with the block hash
`00000000000000004d9b4ef50f0f9d686fd69db2e03af35a100370c64632a983`. Any chain that
doesn't include this block at height 295,000 isn't valid as removing this check

> would be a hard-fork. We note, however, that after version `0.30`, Bitcoin Core

Reviewer 2: Bitcoin Core dropped the leading zero a few years ago:
```suggestion
would be a hard-fork. We note, however, that after v30, Bitcoin Core
```

will remove the checkpoints[^1], as they are not needed anymore to prevent attacks
against nodes during Initial Block Download. This is effectively a hard-fork,
that will probably never actually happen, however.

Block 1,983,702 is the first block that Utreexo nodes would be in danger of a
consensus failure due to the inability to perform the BIP-0030 checks if someone were

> to reuse coinbase transaction from block 164,384. However, this block will happen in roughly

Reviewer 2: ```suggestion
to reuse the coinbase transaction from block 164,384. However, this block will happen in roughly
```

### Historical BIP-0030 violations

There were two UTXOs that were overwritten by repeated transactions:
`e3bf3d07d4b0375638d5f1db5255fe07ba2c4cb067cd81b84ee974b6585fb468:0` at block height 91,722
`d5d27987d2a3dfc724e359870c6644b40e497bdc0589a033220fe15429d88599:0` at block height 91,812

Since the leaf hashes that are committed to the Utreexo accumulator commit to
the block hash as well, all the leaf hashes are unique and the two historical
violations do not happen with how the UTXO set is represented with the Utreexo
accumulator. To be consensus compatible with clients that retain only the second
occurrences of these outputs, the leaves representing the corresponding first UTXOs in the Utreexo accumulator
are hardcoded as unspendable.

These two leaf hashes encoded in hex string are:

 1. `84b3af0783b410b4564c5d1f361868559f7cf77cfc65ce2be951210357022fe3`
 2. `bc6b4bf7cebbd33a18d6b0fe1f8ecc7aa5403083c39ee343b985d51fd0295ad8`

(1) represents the UTXO created at block height 91,722 and (2) represents the
UTXO created at block height 91,812.

## Rationale

**Why use the Utreexo accumulator to keep track of UTXOs instead of a key-value database like leveldb?**

> There's two main advantages to using the Utreexo accumulator instead of a key-value database like leveldb:

Reviewer: ```suggestion
There are two main advantages to using the Utreexo accumulator instead of a key-value database like leveldb:
```

 1. Puts a cap on the UTXO set growth.

>  3. Performance gains with the elimination of random reads/writes.

Reviewer: Renders right of course, but still:

```suggestion
 2. Performance gains with the elimination of random reads/writes.
```

### Puts a cap on the UTXO set growth

The UTXO set is the collection of unspent transaction outputs and is used to verify blocks.
As the amount of Bitcoin users grow, the UTXO set grows with it.

The UTXO set is currently around 10GB in 2025 and with pruning that's all it takes to maintain a full node.
However, as the UTXO set grows, the disk storage requirement will grow along with it and increase the barrier to running a full node.

> Currently, the UTXO set size is $O(log(N))$ where $N$ is the number of UTXOs.

Reviewer 2: I think this is meant to say:
```suggestion
Currently, the UTXO set size is $O(N)$ where $N$ is the number of UTXOs.
```

> By utilizing the Utreexo accumulator, we're able to cap the UTXO set growth at $O(log_2(N))$.

Reviewer: Given that you don’t store the UTXO set, but an accumulator that commits to the UTXO set, perhaps these two sentences should be amended?

 3. The CPU speed that determines how fast the signatures will be validated.

During the default Initial Block Download with assumevalid on, the slowest operation is (2).
This is because fetching and storing the UTXOs are random disk reads and writes, the slowest part of a disk operation.

However when a node is utilizing the Utreexo accumulator, nothing needs to be fetched from the disk.
Instead, the random disk reads and writes are replaced with Utreexo verification which are just hash calculations.

## Reference Implementation

[Utreexod](https://github.com/utreexo/utreexod): A full node implementation with Utreexo support, written in Golang.

[Floresta](https://github.com/vinteumorg/floresta): A lightweight Utreexo client, written in Rust.

## Backward Compatibility

Utreexo nodes are fully backwards compatible with current nodes as they will follow the same chain tip as the current nodes.
Similarly, Utreexo nodes will only consider currently valid transactions for mempool acceptance.

## Acknowledgements

We thank BOB Spaces for lending the space to draft this BIP.

## References

[^1]: https://groups.google.com/g/bitcoindev/c/qyId8Yto45M

> [^2]: https://delvingbitcoin.org/t/great-consensus-cleanup-revival/710

Reviewer 4: nit: i think it's preferable to link to BIP 54 directly nowadays
